alf/algorithms/rlpd_algorithm.py

```diff
@@ -0,0 +1,213 @@
+        """
+        super().__init__(
+            observation_spec=observation_spec,
+            action_spec=action_spec,
+            reward_spec=reward_spec,
+            actor_network_cls=actor_network_cls,
+            critic_network_cls=critic_network_cls,
+            repr_alg_ctor=repr_alg_ctor,
+            reward_weights=reward_weights,
+            train_eps_greedy=train_eps_greedy,
+            epsilon_greedy=epsilon_greedy,
+            use_entropy_reward=use_entropy_reward,
+            normalize_entropy_reward=normalize_entropy_reward,
+            calculate_priority=calculate_priority,
+            num_critic_replicas=num_critic_replicas,
+            env=env,
+            config=config,
+            critic_loss_ctor=critic_loss_ctor,
+            target_entropy=target_entropy,
+            prior_actor_ctor=prior_actor_ctor,
+            target_kld_per_dim=target_kld_per_dim,
+            initial_log_alpha=initial_log_alpha,
+            max_log_alpha=max_log_alpha,
+            target_update_tau=target_update_tau,
+            target_update_period=target_update_period,
+            parameter_reset_period=parameter_reset_period,
+            dqda_clipping=dqda_clipping,
+            actor_optimizer=actor_optimizer,
+            critic_optimizer=critic_optimizer,
+            alpha_optimizer=alpha_optimizer,
+            checkpoint=checkpoint,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        assert self._act_type == ActionType.Continuous, (
+            "RLPD algorithm only supports continuous action spaces.")
+
+        self._num_critic_targets = num_critic_targets
+
+    def _compute_critics(self,
+                         critic_net,
+                         observation,
+                         action,
+                         critics_state,
+                         replica_consensus='mean',
+                         sample_subset=False,
+                         apply_reward_weights=True):
+        observation = (observation, action)
+        # critics shape [B, replicas]
+        critics, critics_state = critic_net(observation, state=critics_state)
+
+        # For multi-dim reward, do
+        # [B, replicas * reward_dim] -> [B, replicas, reward_dim]
+        # For scalar reward, do nothing
+        if self.has_multidim_reward():
+            remaining_shape = critics.shape[2:]
+            critics = critics.reshape(-1, self._num_critic_replicas,
+                                      *self._reward_spec.shape,
+                                      *remaining_shape)
+            if self._act_type == ActionType.Discrete:
+                # permute: [B, replicas, reward_dim, num_actions]
+                #       -> [B, replicas, num_actions, reward_dim]
+                order = [0, 1, -1] + list(
+                    range(2, 2 + len(self._reward_spec.shape)))
+                critics = critics.permute(*order)
+
+        if sample_subset:
+            critics = critics[:,
```

Comment:
when ``self._num_critic_targets`` and ``num_critic_replicas`` are equal, the `randperm`` is not necessary and can be removed to save computation.